### PR TITLE
[Frontend] [Torch] [ONNX] GRU layer

### PR DESCRIPTION
**Pull request by vvchernov**
GRU cell was unified and implemented in common.py. It is used by pytorch and ONNX frontend of TVM. Critical fix was done on ONNX GRU implementation side. Performance tests for different modification of GRU before and after new implementation were carried out. Onnxruntime device is “CPU”. Desktop AMD Ryzen 5 3600 6-Core Processor is used. TVM target is “llvm -mcpu=k6”. The results are collected in the table:

Table 1. Average time per run (microsec) for 10000 runs. The following parameters are used (small input size): with biases = True, batch first = True, linear before reset = True, feature size = 5, hidden size = 10, number of stacked layers = 2, sequence length = 3, batch size = 1, trials number = 100.

| Frontend name/GRU type  |   uni  |    b    |     s     |    sb    | 
| :-----------------------------| :-----:|:------:|:-------:|:-------:|
|Onnx                                    | 23.3  | 48.6   | 46.4   | 100.8   |
|Onnx dev                             | 19.0  | 39.2   | 36.0   |  79.8   |
|Onnx tuned                          |  3.44 | 4.93   |   5.4   |  9.21   |
|Onnx dev tuned                   |  2.9   | 4.1     |  4.42   |  7.48   |
|Pytorch dev                          |  9.49  | 16.17 | 18.87 |  33.3   |
|Pytorch dev tuned               |  2.74  | 3.93   | 3.99   | 6.61   |
|Onnxruntime                       | 13.06  | 16.89 | 19.2  |  25.9   |

Suffix "dev" means implementation in this patch. There are several GRU types: uni – unidirectional, b – bidirectional, s – stacked (2 layers are used in the tests), sb - stacked bidirectional. Compiled by TVM pytorch GRU is faster than tuned ONNX GRU and onnxruntime! Tuned pytorch is strongly quicker. 

Lowlights:
1. using of TVM target “llvm -mcpu=core-avx2” for test machine led to issues related to accuracy after tuning and worse performance.
2. Onnx GRU has issues related wrong implementation but unit test checks accuracy and does not catch it. Need more detail investigation, potentially it can be big problem for other ops.

@masahi please review

### CODE REVIEW

File: python/tvm/relay/frontend/common.py [hunk 1]
```diff
@@ -658,6 +658,92 @@ def unbind(data, axis=0):
...
+    b_inp=None,
+    b_hid=None,
+    rz_act=_op.sigmoid,
+    n_act=_op.tanh,
+    backwards=False,
+    linear_before_reset=True,
+):
+    """
+    Common implementation of GRU cell for all frontends of TVM
+    TODO(vvchernov): currently it is used by pytorch. Extend for other frontends
```
Comment (masahi) [inline python/tvm/relay/frontend/common.py:675]: Also used by ONNX?

File: python/tvm/relay/frontend/common.py [hunk 2]
```diff
@@ -658,6 +658,92 @@ def unbind(data, axis=0):
...
+    result : List[relay.Expr], relay.Expr, relay.Expr
+        The sequence of computed result, final hidden and cell state
+    """
+
+    outputs_list = []
+    for x_t in input_seqs if not backwards else reversed(input_seqs):
+        xwt = _op.nn.dense(x_t, w_inp)
+        if linear_before_reset:
+            hwt = _op.nn.dense(hidden_state, w_hid)
+            # TODO(vvchernov): It is assumed that both bias are or not
```
Comment (masahi) [inline python/tvm/relay/frontend/common.py:712]: What do you mean by `both bias are or not`?

File: python/tvm/relay/frontend/common.py [hunk 3]
```diff
@@ -658,6 +658,92 @@ def unbind(data, axis=0):
...
+            h_r, h_z, h_n = _op.split(hwt, 3, axis=-1)
+            r_gate = rz_act(i_r + h_r)
+            z_gate = rz_act(i_z + h_z)
+            n_gate = n_act(i_n + r_gate * h_n)
+        else:
+            i_r, i_z, i_n = _op.split(xwt, 3, axis=1)
+            w_hr, w_hz, w_hn = _op.split(w_hid, 3, axis=0)
+            r_gate = i_r + _op.nn.dense(hidden_state, w_hr)
+            z_gate = i_z + _op.nn.dense(hidden_state, w_hz)
+            # TODO(vvchernov): It is assumed that both bias are or not
```
Comment (masahi) [inline python/tvm/relay/frontend/common.py:726]: Fix this comment too.

**Comment by masahi** [comment]
@vvchernov Please add PyTorch GRU tests.

**Comment by vvchernov** [comment]
Seems that https://github.com/apache/tvm/blob/main/tests/python/frontend/pytorch/test_forward.py#L4105 duplicates custom lstm test (pytest finds it by its-self). My lstm and gru tests in test_rnns.py but I can transfer them to test_forward.py if need.

File: python/tvm/relay/frontend/common.py (hunk 1 above)
Comment (vvchernov) [reply python/tvm/relay/frontend/common.py]: done

File: python/tvm/relay/frontend/common.py (hunk 2 above)
Comment (vvchernov) [reply python/tvm/relay/frontend/common.py]: fixed

File: python/tvm/relay/frontend/common.py (hunk 3 above)
Comment (vvchernov) [reply python/tvm/relay/frontend/common.py]: fixed

File: tests/python/frontend/pytorch/test_rnns.py [hunk 4]
```diff
@@ -173,6 +253,115 @@ def compare(input, gold_data, rtol=1e-5, atol=1e-5):
...
+                # Get golden output from original model
+                input_hidden_shape = (hidden_layers_num, batch_size, gru_hidden_size)
+                dummy_input, input_shape = model.get_dummy_input()
+                golden_output_batch = model.forward(dummy_input.to(device)).detach().cpu().numpy()
+
+                dtype = "float32"
+                h_zeros = np.zeros(input_hidden_shape, dtype=dtype)
+
+                tvm_output = None
+                for format in ["ts"]:  # ["ts", "onnx"]:
```
Comment (masahi) [inline tests/python/frontend/pytorch/test_rnns.py:313]: This looks like copy paste form the LSTM test. Remove onnx stuff, or use common function for testing both GRU or LSTM.

**Comment by masahi** [comment]
Please try removing duplication between GRU and LSTM tests. For example, I don't want to see 
```
    for batch_first in (True, False):
        for use_bias in (True, False):
            for rnd_weights in [True]:  # (True, False):
```
repeated.

File: tests/python/frontend/pytorch/test_rnns.py (hunk 4 above)
Comment (vvchernov) [reply tests/python/frontend/pytorch/test_rnns.py]: done. Tests were unified

**Review by masahi — APPROVED** [review]

**Comment by masahi** [comment]
Thanks @vvchernov
